core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +98,164 @@ where
             hasher,
             hashes,
             binary_tree_size,
+            head_index: 0,
+            left_cache: vec![],
         }
     }
 
+    /// Returns `true` if the tree is empty.
+    pub fn is_empty(&self) -> bool {
+        self.head_index == 0 && self.hashes.is_empty()
+    }
+
     /// Returns the root hash of this tree.
-    /// # Panics
-    /// Will panic if the constant below is invalid.
-    pub fn merkle_root(self) -> H256 {
+    pub fn merkle_root(&self) -> H256 {
         if self.hashes.is_empty() {
             let depth = tree_depth_by_size(self.binary_tree_size);
-            self.hasher.empty_subtree_hash(depth)
+            if self.head_index == 0 {
+                self.hasher.empty_subtree_hash(depth)
+            } else {
+                self.left_cache[depth]
+            }
         } else {
-            self.compute_merkle_root_and_path(0, None)
+            self.compute_merkle_root_and_path(0, None, None)
         }
     }
 
     /// Returns the root hash and the Merkle proof for a leaf with the specified 0-based `index`.
-    pub fn merkle_root_and_path(self, index: usize) -> (H256, Vec<H256>) {
-        let mut merkle_path = vec![];
-        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut merkle_path));
-        (root_hash, merkle_path)
+    /// `index` is relative to the leftmost uncached leaf.
+    pub fn merkle_root_and_path(&self, index: usize) -> (H256, Vec<H256>) {
+        let (mut left_path, mut right_path) = (vec![], vec![]);
+        let root_hash =
+            self.compute_merkle_root_and_path(index, Some((&mut left_path, &mut right_path)), None);
+        (root_hash, right_path)
```

Comment:
- `left_path` looks unused here. It may make sense to pass `left_path` and `right_path` as separate `Option<&mut _>` values.
- Bikeshedding: IMO, "start path" / "end path" naming would be better because "left" can also be interpreted as a subset of some Merkle path to the left of something.